stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOMStorage.java

```diff
@@ -2,20 +2,197 @@
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.os.Build;
+import com.facebook.stetho.inspector.console.CLog;
+import com.facebook.stetho.inspector.domstorage.DOMStoragePeerManager;
+import com.facebook.stetho.inspector.domstorage.SharedPreferencesHelper;
 import com.facebook.stetho.inspector.jsonrpc.JsonRpcPeer;
+import com.facebook.stetho.inspector.jsonrpc.JsonRpcResult;
 import com.facebook.stetho.inspector.protocol.ChromeDevtoolsDomain;
 import com.facebook.stetho.inspector.protocol.ChromeDevtoolsMethod;
-
+import com.facebook.stetho.json.ObjectMapper;
+import com.facebook.stetho.json.annotation.JsonProperty;
+import org.json.JSONException;
 import org.json.JSONObject;
 
+import java.util.*;
+
 public class DOMStorage implements ChromeDevtoolsDomain {
-  public DOMStorage() {
+  private final Context mContext;
+  private final DOMStoragePeerManager mDOMStoragePeerManager;
+  private final ObjectMapper mObjectMapper = new ObjectMapper();
+
+  public DOMStorage(Context context) {
+    mContext = context;
+    mDOMStoragePeerManager = new DOMStoragePeerManager(context);
   }
 
   @ChromeDevtoolsMethod
   public void enable(JsonRpcPeer peer, JSONObject params) {
+    mDOMStoragePeerManager.addPeer(peer);
   }
 
+  @ChromeDevtoolsMethod
   public void disable(JsonRpcPeer peer, JSONObject params) {
+    mDOMStoragePeerManager.removePeer(peer);
+  }
+
+  @ChromeDevtoolsMethod
+  public JsonRpcResult getDOMStorageItems(JsonRpcPeer peer, JSONObject params)
+      throws JSONException {
+    StorageId storage = mObjectMapper.convertValue(
+        params.getJSONObject("storageId"),
+        StorageId.class);
+
+    ArrayList<List<String>> entries = new ArrayList<List<String>>();
+    String prefTag = storage.securityOrigin;
+    if (storage.isLocalStorage) {
+      SharedPreferences prefs = mContext.getSharedPreferences(prefTag, Context.MODE_PRIVATE);
+      for (Map.Entry<String, ?> prefsEntry : prefs.getAll().entrySet()) {
+        ArrayList<String> entry = new ArrayList<String>(2);
+        entry.add(prefsEntry.getKey());
+        entry.add(SharedPreferencesHelper.valueToString(prefsEntry.getValue()));
+        entries.add(entry);
+      }
+    }
+
+    GetDOMStorageItemsResult result = new GetDOMStorageItemsResult();
+    result.entries = entries;
+    return result;
+  }
+
+  @ChromeDevtoolsMethod
+  public void setDOMStorageItem(JsonRpcPeer peer, JSONObject params) throws JSONException {
+    StorageId storage = mObjectMapper.convertValue(
+        params.getJSONObject("storageId"),
+        StorageId.class);
+    String key = params.getString("key");
+    String value = params.getString("value");
+
+    if (storage.isLocalStorage) {
+      SharedPreferences prefs = mContext.getSharedPreferences(
+          storage.securityOrigin,
+          Context.MODE_PRIVATE);
+      Object existingValue = prefs.getAll().get(key);
+      SharedPreferences.Editor editor = prefs.edit();
```

Comment:
this should be in the `else` block. Why would we call `apply` if we know we arent going to change it.